releasing/plugin-commands-publishing/src/pack.ts

```diff
@@ -63,38 +69,106 @@ export function help (): string {
             description: 'Customizes the output path for the tarball. Use `%s` and `%v` to include the package name and version, e.g., `%s.tgz` or `some-dir/%s-%v.tgz`. By default, the tarball is saved in the current working directory with the name `<package-name>-<version>.tgz`.',
             name: '--out <path>',
           },
+          {
+            description: 'Pack all packages from the workspace',
+            name: '--recursive',
+            shortAlias: '-r',
+          },
         ],
       },
+      FILTERING,
     ],
   })
 }
 
-export type PackOptions = Pick<UniversalOptions, 'dir'> & Pick<Config, 'catalogs' | 'ignoreScripts' | 'rawConfig' | 'embedReadme' | 'packGzipLevel' | 'nodeLinker'> & Partial<Pick<Config, 'extraBinPaths' | 'extraEnv'>> & {
+export type PackOptions = Pick<UniversalOptions, 'dir'> & Pick<Config, 'catalogs'
+| 'ignoreScripts'
+| 'rawConfig'
+| 'embedReadme'
+| 'packGzipLevel'
+| 'nodeLinker'
+> & Partial<Pick<Config, 'extraBinPaths'
+| 'extraEnv'
+| 'recursive'
+| 'selectedProjectsGraph'
+| 'workspaceConcurrency'
+| 'workspaceDir'
+>> & {
   argv: {
     original: string[]
   }
   engineStrict?: boolean
   packDestination?: string
   out?: string
-  workspaceDir?: string
   json?: boolean
+  unicode?: boolean
+}
+
+export interface PackResultJson {
+  name: string
+  version: string
+  filename: string
+  files: Array<{ path: string }>
 }
 
 export async function handler (opts: PackOptions): Promise<string> {
-  const { publishedManifest, tarballPath, contents } = await api(opts)
+  const packedPackages: PackResultJson[] = []
+
+  if (opts.recursive) {
+    const selectedProjectsGraph = opts.selectedProjectsGraph as ProjectsGraph
+    const pkgs = Object.values(selectedProjectsGraph).map((wsPkg) => wsPkg.package)
+    const pkgsToPack = await pFilter(pkgs, pkg => Boolean(pkg.manifest.name && pkg.manifest.version))
+
+    const packedPkgDirs = new Set<ProjectRootDir>(pkgsToPack.map(({ rootDir }) => rootDir))
+
+    if (packedPkgDirs.size === 0) {
+      logger.info({
+        message: 'There are no packages that should be packed',
+        prefix: opts.dir,
+      })
+    }
+
+    const chunks = sortPackages(selectedProjectsGraph)
+
+    const limitPack = pLimit(opts.workspaceConcurrency ?? 4)
+    for (const chunk of chunks) {
+      // eslint-disable-next-line no-await-in-loop
+      await Promise.all(chunk.map(pkgDir =>
+        limitPack(async () => {
+          if (!packedPkgDirs.has(pkgDir)) return
+          const pkg = selectedProjectsGraph[pkgDir].package
+          const packResult = await api({
+            ...opts,
+            dir: pkg.rootDir,
+            // set the packDestination to the current dir if out and packDestination are not set
+            out: opts.out ? path.join(opts.dir, opts.out) : undefined,
+            packDestination: !opts.out ? path.join(opts.dir, opts.packDestination ?? '.') : undefined,
```

Comment:
The packages will be written relative to the current working directory? This is not where the packages are saved currently with non-recursive pack. It is saved relative to the package's directory
```suggestion
            out: opts.out ? path.join(pkg.rootDir, opts.out) : undefined,
            packDestination: !opts.out ? path.join(pkg.rootDir, opts.packDestination ?? '.') : undefined,
```

Reply:
No. If we need to support relative paths, we should make it relative to the cwd, not the workspaceDir. 

in `<root>/packages/a` run `pnpm pack -F a --pack-destination=./out`
relative to cwd: `<root>/packages/a/out/a.tgz`
relative to workspaceDir: `<root>/out/a.tgz` (this is not in line with common practice)



in `<root>` run `pnpm pack -F a --pack-destination=./out`
`<root>/out/a.tgz`

Reply:
https://github.com/pnpm/npm-conf/pull/16

I create a new pull request for supporting get pack-destination from npmrc

Reply:
Added `pack-destination` to the config package https://github.com/pnpm/pnpm/pull/9477/commits/2dc01b3b351b77e43738f136c94486aa4aa6ca38